dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tei/TeiQueryController.java

```diff
@@ -127,6 +134,32 @@ Grid getGrid(
         return teiAnalyticsQueryService.getGrid( params, commonQueryRequest );
     }
 
+    @PreAuthorize( "hasRole('ALL') or hasRole('F_PERFORM_ANALYTICS_EXPLAIN')" )
+    @ResponseBody
+    @GetMapping( value = "query/{trackedEntityType}" + EXPLAIN_PATH, produces = { APPLICATION_JSON_VALUE,
+        "application/javascript" } )
+    Grid getGridExplain(
+        @PathVariable String trackedEntityType,
+        TeiQueryRequest teiQueryRequest,
+        CommonQueryRequest commonQueryRequest,
+        HttpServletResponse response )
+    {
+        QueryRequest<TeiQueryRequest> queryRequest = QueryRequest.<TeiQueryRequest> builder()
+            .request( teiQueryRequestProcessor.process(
+                teiQueryRequest.withTrackedEntityType( trackedEntityType ) ) )
+            .commonQueryRequest( commonQueryRequestProcessor.process( commonQueryRequest ) )
+            .build();
+
+        commonQueryRequestValidator.validate( queryRequest.getCommonQueryRequest() );
+        teiQueryRequestValidator.validate( queryRequest );
+
+        contextUtils.configureResponse( response, CONTENT_TYPE_JSON, RESPECT_SYSTEM_SETTING );
+
+        TeiQueryParams params = mapper.map( queryRequest );
+
+        return teiAnalyticsQueryService.getGridExplain( params );
```

Comment:
we might extract this method in a common private one, and invoke from each controller's methods